diff --git a/src/meson.build b/src/meson.build
index 594e808..588830e 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -76,7 +76,7 @@ paper_deps = [
   dependency('gtk4'),
   dependency('gio-2.0'),
   dependency('gee-0.8'),
-  dependency('libadwaita-1', version: '1.2.0', required: true),
+  dependency('libadwaita-1', version: '>=1.2.0', required: true),
   dependency('gtksourceview-5'),
   m_dep,
 ]
