Index: src/conf.h
===================================================================
--- src/conf.h	(revision 5437)
+++ src/conf.h	(working copy)
@@ -45,6 +45,7 @@
 #define DEFAULT_MAX_ITEMS		"/apps/liferea/maxitemcount"
 #define DEFAULT_UPDATE_INTERVAL		"/apps/liferea/default-update-interval"
 #define STARTUP_FEED_ACTION		"/apps/liferea/startup_feed_action"
+#define ENABLE_FEED_URL_UPDATE		"/apps/liferea/enable_feed_url_update"
 
 /* folder handling settings */
 #define FOLDER_DISPLAY_MODE		"/apps/liferea/folder-display-mode"
Index: src/subscription.c
===================================================================
--- src/subscription.c	(revision 5437)
+++ src/subscription.c	(working copy)
@@ -199,12 +199,14 @@
 	nodePtr		node = subscription->node;
 	gboolean	processing = FALSE;
 	GTimeVal	now;
+	gboolean	enable_url_update;
 
 	/* 1. preprocessing */
 
 	g_assert (subscription->updateJob);
 	/* update the subscription URL on permanent redirects */
-	if (result->source && !g_str_equal (result->source, subscription->updateJob->request->source)) {
+	conf_get_bool_value (ENABLE_FEED_URL_UPDATE, &enable_url_update);
+	if (enable_url_update && result->source && !g_str_equal (result->source, subscription->updateJob->request->source)) {
 		debug2 (DEBUG_UPDATE, "The URL of \"%s\" has changed permanently and was updated with \"%s\"", node_get_title(node), result->source);
 		subscription_set_source (subscription, result->source);
 		liferea_shell_set_status_bar (_("The URL of \"%s\" has changed permanently and was updated"), node_get_title(node));
Index: src/ui/ui_prefs.c
===================================================================
--- src/ui/ui_prefs.c	(revision 5437)
+++ src/ui/ui_prefs.c	(working copy)
@@ -451,6 +451,12 @@
 	liferea_shell_update_toolbar ();
 }
 
+void
+on_update_feed_url_value_changed (GtkToggleButton *button, gpointer user_data)
+{
+	conf_set_bool_value (ENABLE_FEED_URL_UPDATE, gtk_toggle_button_get_active (button));
+}
+
 /*------------------------------------------------------------------------------*/
 /* preferences dialog callbacks 						*/
 /*------------------------------------------------------------------------------*/
@@ -486,6 +492,7 @@
 	gboolean		start_in_tray, disable_toolbar;
 	gchar			*proxy_host, *proxy_user, *proxy_passwd;
 	gchar			*browser_command, *enclosure_download_path;
+	gboolean 		enable_feed_url_update;
 	
 	if (!prefdialog) {
 		prefdialog = liferea_dialog_new (NULL, "prefdialog");
@@ -560,6 +567,11 @@
 		gtk_spin_button_set_value (GTK_SPIN_BUTTON (widget), tmp);
 		g_signal_connect (G_OBJECT (widget), "changed", G_CALLBACK (on_default_update_interval_value_changed), NULL);
 
+		conf_get_bool_value (ENABLE_FEED_URL_UPDATE, &enable_feed_url_update);
+		widget = liferea_dialog_lookup (prefdialog, "updateFeedURLCheckButton");
+		gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (widget), enable_feed_url_update);
+		g_signal_connect (G_OBJECT (widget), "changed", G_CALLBACK (on_update_feed_url_value_changed), NULL);
+
 		/* ================== panel 2 "folders" ==================== */
 
 		g_signal_connect(GTK_OBJECT(liferea_dialog_lookup(prefdialog, "updateAllFavicons")), "clicked", G_CALLBACK(on_updateallfavicons_clicked), NULL);
