diff --git a/setup.py b/setup.py
index 1ae9df7..9b7de52 100644
--- a/setup.py
+++ b/setup.py
@@ -1,5 +1,6 @@
 from distutils.core import setup
 from distutils.extension import Extension
+from distutils.sysconfig import get_python_lib
 
 import sys, os.path
 
@@ -8,7 +9,7 @@ import sys, os.path
 #if distutils.sysconfig.get_config_var('CC').startswith("gcc"):
 #    pyrex_compile_options = []
 #else:
-pyrex_compile_options = []
+pyrex_compile_options = ["-I", os.path.join(get_python_lib(), "numpy/core/include")]
 
 if sys.platform == "win32" and len(sys.argv) < 2:
     sys.argv[1:] = ["bdist_wininst"]
