diff --git a/nm-applet.conf b/nm-applet.conf
index ebd7bc6..48c623e 100644
--- a/nm-applet.conf
+++ b/nm-applet.conf
@@ -7,7 +7,7 @@
 
                 <allow send_destination="org.freedesktop.NetworkManagerUserSettings"/>
         </policy>
-        <policy at_console="true">
+        <policy group="plugdev">
                 <allow own="org.freedesktop.NetworkManagerUserSettings"/>
 
                 <allow send_destination="org.freedesktop.NetworkManagerUserSettings"/>
