diff --git a/httplib2-0.5.0-2.6/httplib2/__init__.py b/httplib2-0.5.0-2.6/httplib2/__init__.py
index 56c018b..e4c6924 100755
--- a/httplib2-0.5.0-2.6/httplib2/__init__.py
+++ b/httplib2-0.5.0-2.6/httplib2/__init__.py
@@ -847,6 +847,7 @@ the same interface as FileCache."""
     def _conn_request(self, conn, request_uri, method, body, headers):
         for i in range(2):
             try:
+                conn.connect()
                 conn.request(method, request_uri, body, headers)
             except socket.gaierror:
                 conn.close()
diff --git a/httplib2-0.5.0-3.1/httplib2/__init__.py b/httplib2-0.5.0-3.1/httplib2/__init__.py
index fac2b90..4080b74 100755
--- a/httplib2-0.5.0-3.1/httplib2/__init__.py
+++ b/httplib2-0.5.0-3.1/httplib2/__init__.py
@@ -832,6 +832,7 @@ the same interface as FileCache."""
     def _conn_request(self, conn, request_uri, method, body, headers):
         for i in range(2):
             try:
+                conn.connect()
                 conn.request(method, request_uri, body, headers)
             except socket.gaierror:
                 conn.close()
diff --git a/httplib2-0.5.0/httplib2/__init__.py b/httplib2-0.5.0/httplib2/__init__.py
index 56c018b..e4c6924 100755
--- a/httplib2-0.5.0/httplib2/__init__.py
+++ b/httplib2-0.5.0/httplib2/__init__.py
@@ -847,6 +847,7 @@ the same interface as FileCache."""
     def _conn_request(self, conn, request_uri, method, body, headers):
         for i in range(2):
             try:
+                conn.connect()
                 conn.request(method, request_uri, body, headers)
             except socket.gaierror:
                 conn.close()
diff --git a/httplib2-python3-0.5.0/httplib2/__init__.py b/httplib2-python3-0.5.0/httplib2/__init__.py
index fac2b90..4080b74 100755
--- a/httplib2-python3-0.5.0/httplib2/__init__.py
+++ b/httplib2-python3-0.5.0/httplib2/__init__.py
@@ -832,6 +832,7 @@ the same interface as FileCache."""
     def _conn_request(self, conn, request_uri, method, body, headers):
         for i in range(2):
             try:
+                conn.connect()
                 conn.request(method, request_uri, body, headers)
             except socket.gaierror:
                 conn.close()
