diff --git a/libgiggle-git/Makefile.am b/libgiggle-git/Makefile.am
index 184aec1..92f4f7f 100644
--- a/libgiggle-git/Makefile.am
+++ b/libgiggle-git/Makefile.am
@@ -63,7 +63,7 @@ libgiggle_git_HEADERS = \
 	$(NULL)
 
 libgiggle_git_la_LIBADD = \
-	$(GIGGLE_LIBS) ../libgiggle/libgiggle.la
+	../libgiggle/libgiggle.la $(GIGGLE_LIBS)
 libgiggle_git_la_LDFLAGS = \
 	$(AM_LDFLAGS) \
 	-release $(PACKAGE_VERSION) \
