diff -uNr netspeed_applet-0.14.orig/src/backend.c netspeed_applet-0.14/src/backend.c
--- netspeed_applet-0.14.orig/src/backend.c	2006-08-06 19:21:43.000000000 +0100
+++ netspeed_applet-0.14/src/backend.c	2007-04-27 23:16:05.000000000 +0100
@@ -25,7 +25,7 @@
 gboolean
 is_no_dummy_device(const char* device)
 {
-	return !g_str_has_prefix(device, "lo") && !g_str_has_prefix(device, "dummy");
+	return !g_str_has_prefix(device, "lo") && !g_str_has_prefix(device, "wmaster") && !g_str_has_prefix(device, "dummy") && !g_str_has_prefix(device, "cipsec") && !g_str_has_prefix(device, "wifi") && !g_str_has_prefix(device, "vmnet");
 }
 
 
