diff -uNr a/lib/libokbase2.pc.in b/lib/libokbase2.pc.in
--- a/lib/libokbase2.pc.in	2007-06-17 13:59:16.000000000 +0100
+++ b/lib/libokbase2.pc.in	2007-06-17 13:59:28.000000000 +0100
@@ -1,4 +1,4 @@
-prefix=/usr/local
+prefix=/usr
 exec_prefix=${prefix}
 libdir=${exec_prefix}/lib
 includedir=${prefix}/include
diff -uNr a/lib/libsyncapi.pc.in b/lib/libsyncapi.pc.in
--- a/lib/libsyncapi.pc.in	2007-06-17 13:59:16.000000000 +0100
+++ b/lib/libsyncapi.pc.in	2007-06-17 13:59:37.000000000 +0100
@@ -1,4 +1,4 @@
-prefix=/usr/local
+prefix=/usr
 exec_prefix=${prefix}
 libdir=${exec_prefix}/lib
 includedir=${prefix}/include
