diff --git a/aux/broctl/CMakeLists.txt b/aux/broctl/CMakeLists.txt
index a7b98fa..b2428ce 100644
--- a/aux/broctl/CMakeLists.txt
+++ b/aux/broctl/CMakeLists.txt
@@ -206,19 +206,6 @@ endif ()
 configure_file(etc/broctl.cfg.in
                ${CMAKE_CURRENT_BINARY_DIR}/etc/broctl.cfg)
 
-InstallPackageConfigFile(
-    ${CMAKE_CURRENT_BINARY_DIR}/etc/broctl.cfg
-    ${ETC}
-    broctl.cfg)
-InstallPackageConfigFile(
-    ${CMAKE_CURRENT_SOURCE_DIR}/etc/networks.cfg
-    ${ETC}
-    networks.cfg)
-InstallPackageConfigFile(
-    ${CMAKE_CURRENT_SOURCE_DIR}/etc/node.cfg
-    ${ETC}
-    node.cfg)
-
 ########################################################################
 ## Packaging Setup
 
