diff --git a/src/btctlimpl.c b/src/btctlimpl.c
index a3d86c9..91636aa 100644
--- a/src/btctlimpl.c
+++ b/src/btctlimpl.c
@@ -330,7 +330,7 @@ btctl_controller_impl_cmd_scan(BtctlController *bc, GError **err)
 	if (dd>=0) {
 		for (i = 0; i < num_rsp; i++) {
 			memset(name, 0, sizeof(name));
-			if (hci_remote_name(dd, &(info+i)->bdaddr,
+			if (hci_read_remote_name(dd, &(info+i)->bdaddr,
 						sizeof(name), name, 100000) < 0)
 				strcpy(name, "n/a");
 
