diff --git a/gen_compile.sh b/gen_compile.sh
index 3379b61..1c9db7a 100755
--- a/gen_compile.sh
+++ b/gen_compile.sh
@@ -79,6 +79,10 @@ export_utils_args()
 	then
 		export AS="${UTILS_AS}"
 	fi
+	if [ "${UTILS_CROSS_COMPILE}" != '' ]
+	then
+		export CROSS_COMPILE="${UTILS_CROSS_COMPILE}"
+	fi
 }
 
 unset_utils_args()
@@ -99,6 +103,10 @@ unset_utils_args()
 	then
 		unset AS
 	fi
+	if [ "${UTILS_CROSS_COMPILE}" != '' ]
+	then
+		unset CROSS_COMPILE
+	fi
 	reset_args
 }
 
@@ -393,7 +401,7 @@ compile_busybox() {
 		print_info 1 'busybox: >> Copying to cache...'
 		[ -f "${TEMP}/${BUSYBOX_DIR}/busybox" ] ||
 			gen_die 'Busybox executable does not exist!'
-		strip "${TEMP}/${BUSYBOX_DIR}/busybox" ||
+		${UTILS_CROSS_COMPILE}strip "${TEMP}/${BUSYBOX_DIR}/busybox" ||
 			gen_die 'Could not strip busybox binary!'
 		tar -cj -C "${TEMP}/${BUSYBOX_DIR}" -f "${BUSYBOX_BINCACHE}" busybox .config .config.gk_orig ||
 			gen_die 'Could not create the busybox bincache!'
@@ -432,7 +440,7 @@ compile_lvm() {
 
 		cd "${TEMP}/lvm"
 		print_info 1 '      >> Copying to bincache...'
-		strip "sbin/lvm.static" ||
+		${UTILS_CROSS_COMPILE}strip "sbin/lvm.static" ||
 			gen_die 'Could not strip lvm.static!'
 		/bin/tar -cjf "${LVM_BINCACHE}" sbin/lvm.static ||
 			gen_die 'Could not create binary cache'
@@ -511,7 +519,7 @@ compile_device_mapper() {
 		cd "${TEMP}"
 		rm -rf "${TEMP}/device-mapper/man" ||
 			gen_die 'Could not remove manual pages!'
-		strip "${TEMP}/device-mapper/sbin/dmsetup" ||
+		${UTILS_CROSS_COMPILE}strip "${TEMP}/device-mapper/sbin/dmsetup" ||
 			gen_die 'Could not strip dmsetup binary!'
 		/bin/tar -jcpf "${DEVICE_MAPPER_BINCACHE}" device-mapper ||
 			gen_die 'Could not tar up the device-mapper binary!'
@@ -542,7 +550,7 @@ compile_e2fsprogs() {
 		print_info 1 'blkid: >> Copying to cache...'
 		[ -f "${TEMP}/${E2FSPROGS_DIR}/misc/blkid" ] ||
 			gen_die 'Blkid executable does not exist!'
-		strip "${TEMP}/${E2FSPROGS_DIR}/misc/blkid" ||
+		${UTILS_CROSS_COMPILE}strip "${TEMP}/${E2FSPROGS_DIR}/misc/blkid" ||
 			gen_die 'Could not strip blkid binary!'
 		bzip2 "${TEMP}/${E2FSPROGS_DIR}/misc/blkid" ||
 			gen_die 'bzip2 compression of blkid failed!'
@@ -575,7 +583,7 @@ compile_fuse() {
 #		print_info 1 'libfuse: >> Copying to cache...'
 #		[ -f "${TEMP}/${FUSE_DIR}/lib/.libs/libfuse.so" ] ||
 #			gen_die 'libfuse.so does not exist!'
-#		strip "${TEMP}/${FUSE_DIR}/lib/.libs/libfuse.so" ||
+#		${UTILS_CROSS_COMPILE}strip "${TEMP}/${FUSE_DIR}/lib/.libs/libfuse.so" ||
 #			gen_die 'Could not strip libfuse.so!'
 #		cd "${TEMP}/${FUSE_DIR}/lib/.libs"
 #		tar -cjf "${FUSE_BINCACHE}" libfuse*so* ||
@@ -608,7 +616,7 @@ compile_unionfs_fuse() {
 		print_info 1 'unionfs-fuse: >> Copying to cache...'
 		[ -f "${TEMP}/${UNIONFS_FUSE_DIR}/src/unionfs" ] ||
 			gen_die 'unionfs binary does not exist!'
-		strip "${TEMP}/${UNIONFS_FUSE_DIR}/src/unionfs" ||
+		${UTILS_CROSS_COMPILE}strip "${TEMP}/${UNIONFS_FUSE_DIR}/src/unionfs" ||
 			gen_die 'Could not strip unionfs binary!'
 		bzip2 "${TEMP}/${UNIONFS_FUSE_DIR}/src/unionfs" ||
 			gen_die 'bzip2 compression of unionfs binary failed!'
