diff --git a/pango/pango-layout.c b/pango/pango-layout.c
index 4cf00a1..2d5add1 100644
--- a/pango/pango-layout.c
+++ b/pango/pango-layout.c
@@ -1832,12 +1832,12 @@ pango_layout_move_cursor_visually (PangoLayout *layout,
 /**
  * pango_layout_xy_to_index:
  * @layout:    a #PangoLayout
- * @x:         the X offset (in Pango units)
+ * @x: (in):   the X offset (in Pango units)
  *             from the left edge of the layout.
- * @y:         the Y offset (in Pango units)
+ * @y: (in):   the Y offset (in Pango units)
  *             from the top edge of the layout
- * @index_:    location to store calculated byte index
- * @trailing:  location to store a integer indicating where
+ * @index_: (out): location to store calculated byte index
+ * @trailing: (out): location to store a integer indicating where
  *             in the grapheme the user clicked. It will either
  *             be zero, or the number of characters in the
  *             grapheme. 0 represents the trailing edge of the grapheme.
@@ -1937,8 +1937,8 @@ pango_layout_xy_to_index (PangoLayout *layout,
 /**
  * pango_layout_index_to_pos:
  * @layout: a #PangoLayout
- * @index_: byte index within @layout
- * @pos: rectangle in which to store the position of the grapheme
+ * @index_: (in): byte index within @layout
+ * @pos: (out): rectangle in which to store the position of the grapheme
  *
  * Converts from an index within a #PangoLayout to the onscreen position
  * corresponding to the grapheme at that index, which is represented
@@ -2164,9 +2164,9 @@ pango_layout_line_get_char_direction (PangoLayoutLine *layout_line,
 /**
  * pango_layout_get_cursor_pos:
  * @layout: a #PangoLayout
- * @index_: the byte index of the cursor
- * @strong_pos: location to store the strong cursor position (may be %NULL)
- * @weak_pos: location to store the weak cursor position (may be %NULL)
+ * @index_: (in): the byte index of the cursor
+ * @strong_pos: (out) (allow-none): location to store the strong cursor position (may be %NULL)
+ * @weak_pos: (out) (allow-none): location to store the weak cursor position (may be %NULL)
  *
  * Given an index within a layout, determines the positions that of the
  * strong and weak cursors if the insertion point is at that
@@ -2586,9 +2586,9 @@ pango_layout_get_extents_internal (PangoLayout    *layout,
 /**
  * pango_layout_get_extents:
  * @layout:   a #PangoLayout
- * @ink_rect: rectangle used to store the extents of the layout as drawn
+ * @ink_rect: (out) (allow-none): rectangle used to store the extents of the layout as drawn
  *            or %NULL to indicate that the result is not needed.
- * @logical_rect: rectangle used to store the logical extents of the layout
+ * @logical_rect: (out) (allow-none): rectangle used to store the logical extents of the layout
 		 or %NULL to indicate that the result is not needed.
  *
  * Computes the logical and ink extents of @layout. Logical extents
@@ -2614,9 +2614,9 @@ pango_layout_get_extents (PangoLayout    *layout,
 /**
  * pango_layout_get_pixel_extents:
  * @layout:   a #PangoLayout
- * @ink_rect: rectangle used to store the extents of the layout as drawn
+ * @ink_rect: (out) (allow-none): rectangle used to store the extents of the layout as drawn
  *            or %NULL to indicate that the result is not needed.
- * @logical_rect: rectangle used to store the logical extents of the
+ * @logical_rect: (out) (allow-none): rectangle used to store the logical extents of the
  *              layout or %NULL to indicate that the result is not needed.
  *
  * Computes the logical and ink extents of @layout in device units.
@@ -2640,8 +2640,8 @@ pango_layout_get_pixel_extents (PangoLayout *layout,
 /**
  * pango_layout_get_size:
  * @layout: a #PangoLayout
- * @width: location to store the logical width, or %NULL
- * @height: location to store the logical height, or %NULL
+ * @width: (out) (allow-none): location to store the logical width, or %NULL
+ * @height: (out) (allow-none): location to store the logical height, or %NULL
  *
  * Determines the logical width and height of a #PangoLayout
  * in Pango units (device units scaled by %PANGO_SCALE). This
@@ -2665,8 +2665,8 @@ pango_layout_get_size (PangoLayout *layout,
 /**
  * pango_layout_get_pixel_size:
  * @layout: a #PangoLayout
- * @width: location to store the logical width, or %NULL
- * @height: location to store the logical height, or %NULL
+ * @width: (out) (allow-none): location to store the logical width, or %NULL
+ * @height: (out) (allow-none): location to store the logical height, or %NULL
  *
  * Determines the logical width and height of a #PangoLayout
  * in device units. (pango_layout_get_size() returns the width
