diff --git a/Makefile.am b/Makefile.am
index d5e0305..58cae77 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -7,7 +7,6 @@ SUBDIRS = \
 	Alias \
 	AptURL \
 	Archive \
-	Banshee \
 	Bibtex \
 	Cl.ickable \
 	ClawsMail \
@@ -16,10 +15,8 @@ SUBDIRS = \
 	DiskMounter \
 	EOG-Slideshow \
 	Epiphany \
-	Evolution \
 	File \
 	Firefox \
-	Flickr \
 	GNOME-Dictionary \
 	GNOME-Screenshot \
 	GNOME-Session \
diff --git a/configure.ac b/configure.ac
index 9b39979..0207bd2 100644
--- a/configure.ac
+++ b/configure.ac
@@ -30,9 +30,6 @@ SHAMROCK_EXPAND_LIBDIR
 SHAMROCK_EXPAND_BINDIR
 SHAMROCK_EXPAND_DATADIR
 
-PKG_CHECK_MODULES([BANSHEE_INDEXER], [banshee-1-collection-indexer >= 1.4.2])
-PKG_CHECK_MODULES([EVOLUTION_SHARP], [evolution-sharp])
-PKG_CHECK_MODULES([FLICKRNET], [flickrnet])
 PKG_CHECK_MODULES([GCONF_SHARP_20], [gconf-sharp-2.0])
 PKG_CHECK_MODULES([GLADE_SHARP_20], [glade-sharp-2.0])
 PKG_CHECK_MODULES([GLIB_SHARP_20], [glib-sharp-2.0])
@@ -55,9 +52,6 @@ GDATA_SHARP_CALENDAR_LIBS="-r:$top_srcdir/BundledLibraries/Google.GData.Client.d
 GDATA_SHARP_DOCUMENTS_LIBS="-r:$top_srcdir/BundledLibraries/Google.GData.Client.dll -r:$top_srcdir/BundledLibraries/Google.GData.Extensions.dll -r:$top_srcdir/BundledLibraries/Google.GData.Documents.dll"
 GDATA_SHARP_CONTACTS_LIBS="-r:$top_srcdir/BundledLibraries/Google.GData.Client.dll -r:$top_srcdir/BundledLibraries/Google.GData.Extensions.dll -r:$top_srcdir/BundledLibraries/Google.GData.Contacts.dll"
 
-AC_SUBST([BANSHEE_INDEXER_LIBS])
-AC_SUBST([EVOLUTION_SHARP_LIBS])
-AC_SUBST([FLICKRNET_LIBS])
 AC_SUBST([GCONF_SHARP_20_LIBS])
 AC_SUBST([GDATA_SHARP_ACL_LIBS])
 AC_SUBST([GDATA_SHARP_CALENDAR_LIBS])
@@ -112,7 +106,6 @@ AC_CONFIG_FILES([
 Alias/Makefile
 AptURL/Makefile
 Archive/Makefile
-Banshee/Makefile
 Bibtex/Makefile
 Cl.ickable/Makefile
 ClawsMail/Makefile
@@ -121,10 +114,8 @@ Del.icio.us/Makefile
 DiskMounter/Makefile
 EOG-Slideshow/Makefile
 Epiphany/Makefile
-Evolution/Makefile
 File/Makefile
 Firefox/Makefile
-Flickr/Makefile
 GNOME-Dictionary/Makefile
 GNOME-Screenshot/Makefile
 GNOME-Session/Makefile
