diff --git a/pypy/translator/c/gcc/trackgcroot.py b/pypy/translator/c/gcc/trackgcroot.py
index c15d472..f4b08d1 100755
--- a/pypy/translator/c/gcc/trackgcroot.py
+++ b/pypy/translator/c/gcc/trackgcroot.py
@@ -491,6 +491,10 @@ class FunctionGcRootTracker(object):
         'movz', 
         # locked operations should not move GC pointers, at least so far
         'lock',
+        # additional unknown operations
+        'vmul', 'vcvt', 'vucomi', 'vsub', 'vxor', 'vmov', 'vdiv',
+        'vadd', 'vcmp', 'vand', 'vor', 'vunpck', 'vsqrt', 'vmax',
+        'vmin',
         ])
 
     # a partial list is hopefully good enough for now; it's all to support
