diff --git a/src/libnfc.c b/src/libnfc.c
index cdb8524..6285963 100644
--- a/src/libnfc.c
+++ b/src/libnfc.c
@@ -645,7 +645,7 @@ bool nfc_target_init(const dev_info* pdi, byte_t* pbtRx, uint32_t* puiRxBits)
 	
   // Make sure the CRC & parity are handled by the device, this is needed for target_init to work properly
   if (!bCrc) nfc_configure((dev_info*)pdi,DCO_HANDLE_CRC,true);
-  if (!bPar) nfc_configure((dev_info*)pdi,DCO_HANDLE_CRC,true);
+  if (!bPar) nfc_configure((dev_info*)pdi,DCO_HANDLE_PARITY,true);
 
   // Let the PN53X be activated by the RF level detector from power down mode
   if (!pn53x_set_reg(pdi,REG_CIU_TX_AUTO, SYMBOL_INITIAL_RF_ON,0x04)) return false;
@@ -665,7 +665,7 @@ bool nfc_target_init(const dev_info* pdi, byte_t* pbtRx, uint32_t* puiRxBits)
 
   // Restore the CRC & parity setting to the original value (if needed)
   if (!bCrc) nfc_configure((dev_info*)pdi,DCO_HANDLE_CRC,false);
-  if (!bPar) nfc_configure((dev_info*)pdi,DCO_HANDLE_CRC,false);
+  if (!bPar) nfc_configure((dev_info*)pdi,DCO_HANDLE_PARITY,false);
 
   return true;
 }
