diff --git a/mailnag b/mailnag
index 934474e..cfa37e8 100755
--- a/mailnag
+++ b/mailnag
@@ -1,5 +1,5 @@
 #!/bin/bash
-LIB_DIR=./Mailnag
+LIB_DIR=/usr/share/mailnag/Mailnag
 
 config_dir="${XDG_CONFIG_HOME:-$HOME/.config}/mailnag"
 
@@ -18,5 +18,5 @@ fi
 
 rm --force "$config_dir/mailnag.log"
 
-python $LIB_DIR/mailnag.py >> "$config_dir/mailnag.log" 2>&1 &
+python2.7 $LIB_DIR/mailnag.py >> "$config_dir/mailnag.log" 2>&1 &
 
diff --git a/mailnag_config b/mailnag_config
index 0fd379e..6522140 100755
--- a/mailnag_config
+++ b/mailnag_config
@@ -1,7 +1,7 @@
 #!/bin/bash
-LIB_DIR=./Mailnag
+LIB_DIR=/usr/share/mailnag/Mailnag
 
-python $LIB_DIR/config.py
+python2.7 $LIB_DIR/config.py
 
 if [ $? -eq 0 ]; then
 	# Restart mailnag daemon
