diff --git a/Mailnag/common/utils.py b/Mailnag/common/utils.py
index 1717679..05d0e2f 100644
--- a/Mailnag/common/utils.py
+++ b/Mailnag/common/utils.py
@@ -39,7 +39,7 @@ def get_data_file(filename):
 	# Add "./data" in workdir for running from builddir
 	data_paths = []
 	data_paths.append("./data")
-	data_paths.extend(base.load_data_paths(PACKAGE_NAME))
+	data_paths.extend([os.path.join(x, 'data') for x in base.load_data_paths(PACKAGE_NAME)])
 
 	for direc in data_paths:
 		file_path = os.path.join(direc, filename)
