diff --git a/bindings/wscript_build b/bindings/wscript_build
index 36de9c9..1084c2b 100644
--- a/bindings/wscript_build
+++ b/bindings/wscript_build
@@ -4,24 +4,27 @@ import Build
 
 bld.use_the_magic()
 
-Task.simple_task_type('swig_python', '${SWIG} -o ${TGT} -python ${SRC}', before = 'cc')
+Task.simple_task_type('swig_python', '${SWIG} -o ${TGT} -python ${SRC}', before = 'cc', ext_out = ['.c', '.py'])
 
 @TaskGen.extension('.i')
 def process_swig(self, node):
 	cc_node = node.change_ext('.c')
-	self.create_task('swig_python', node, cc_node)
+	py_node = node.change_ext('.py')
+	task = self.create_task('swig_python', node, cc_node)
 	self.allnodes.append(cc_node)
-	
+	self.allnodes.append(py_node)
+
 if 'python' in bld.env.bindings:
 	if 'SWIG' not in bld.env:
 		raise Build.BuildError('Swig is needed to build python bindings')
 	py_bindings = bld(
-			features = ['cc', 'cshlib'],
+			features = ['py', 'cc', 'cshlib'],
 			source   = 'libbls.i',
-			includes = ['/usr/include/python2.6'],	
+			includes = bld.env.CPPPATH_PYEXT,
 			uselib_local = 'bls-%s' % bld.env.LIBBLS_VERSION_NO_PATCH,
 			target   = '_libbls',
-			env = bld.env_of_name('visibility-public').copy()
+			env = bld.env_of_name('visibility-public').copy(),
+			install_path = bld.env.PYTHONDIR
 			)
 	py_bindings.env.shlib_PATTERN = py_bindings.env.shlib_PATTERN.replace('lib%s', '%s')
 
diff --git a/doc/user/wscript_build b/doc/user/wscript_build
index 17e733a..437ad3e 100644
--- a/doc/user/wscript_build
+++ b/doc/user/wscript_build
@@ -3,7 +3,7 @@ import Options
 user_doc = bld(
 	target = 'user_guide.html',
 	source = ['user_guide.txt', 'user_guide.css'],
-	rule = 'rst2html --initial-header-level 2 --stylesheet-path ${SRC[1].abspath()} ${SRC[0].abspath()} > ${TGT}',
+	rule = 'rst2html.py --initial-header-level 2 --stylesheet-path ${SRC[1].abspath()} ${SRC[0].abspath()} > ${TGT}',
 	install_path = '${PREFIX}/share/doc/libbls-%s' % bld.env.LIBBLS_VERSION_NO_PATCH
 )
 bld.install_files('${PREFIX}/share/doc/libbls-${LIBBLS_VERSION_NO_PATCH}', ['user_guide.txt', 'user_guide.css'])
diff --git a/wscript b/wscript
index 4369638..b496758 100644
--- a/wscript
+++ b/wscript
@@ -31,6 +31,7 @@ def set_options(opt):
 	
 def configure(conf):
 	conf.check_tool('compiler_cc')
+	conf.check_tool('python')
 	conf.check_tool('misc')
 	
 	# Check required headers
@@ -89,6 +90,10 @@ def configure(conf):
 	(conf.env.LIBBLS_VERSION_MAJOR, conf.env.LIBBLS_VERSION_MINOR, conf.env.LIBBLS_VERSION_PATCH) = VERSION.split('.')
 	conf.env.LIBBLS_VERSION_NO_PATCH = '%s.%s' % (conf.env.LIBBLS_VERSION_MAJOR, conf.env.LIBBLS_VERSION_MINOR)
 
+	# Check for python
+	conf.check_python_version((2,6,0))
+	conf.check_python_headers()
+
 	# Create a environment with default (public) symbol visibility
 	env = conf.env.copy()
 	env.append_unique('CCFLAGS', '-fvisibility=default'.split())
